examples/cosmos/ics20/bank.qnt

```diff
@@ -0,0 +1,240 @@
+// -*- mode: Bluespec; -*-
+
+/**
+ * A bank module for ICS-020 (Fungible Token Transfer).
+ *
+ * ICS-020 assumes a high-level "bank module" capable of minting, burning, and
+ * transfering tokens. This specification is held as general as possible,
+ * providing only the assumptions required by ICS-020.
+ *
+ * For a possible implementation, consider the Cosmos SDK bank module:
+ * https://docs.cosmos.network/v0.47/modules/bank/
+ *
+ * Thomas Pani, Informal Systems, 2023
+ */
+
+module bank {
+
+  /* ***************************************************************************
+   * TYPES
+   * **************************************************************************/
+
+  type Address = str
+  type Denomination = str
+  type Amount = int
+
+  type TokenBalances = Denomination -> Amount
+  type Accounts = Address -> TokenBalances
+
+  /* ***************************************************************************
+   * FUNCTIONAL LAYER, manipulating balances
+   * **************************************************************************/
+
+  // Get balances of all denominations in the given account `addr`.
+  pure def getBalances(accounts: Accounts, addr: Address) : TokenBalances = {
+    if (accounts.keys().contains(addr)) {
+      accounts.get(addr)
+    } else {
+      Map()
+    }
+  }
+
+  // Get the amount of tokens of denomination `denom` in `balances`.
+  pure def getBalance(balances: TokenBalances, denom: Denomination): Amount = {
+    if (balances.keys().contains(denom)) balances.get(denom) else 0
+  }
+
+  /* Return `accounts`, where `address` is updated to hold
+   * `currentAmount + amount` tokens of denomination `denom`.
+   * 
+   * `currentAmount` refers to the number of tokens of denomination `denom` held
+   * by `address` in `accounts`.
+   *
+   * If there currently is no account under `address`, return accounts including
+   * `address` with exactly `amount`-many `denom` tokens.
+   */
+  pure def manipulateAccount(accounts: Accounts, address: Address,
+                             denom: Denomination, amount: Amount) : Accounts = {
+    val accountCoins = accounts.getBalances(address)
+    val newCoins = accountCoins.put(denom, accountCoins.getBalance(denom) + amount)
```

Comment:
Just curious, the ICS specs do not assume any bound on the amounts? Or maybe they assume that there is a bound, but it is different for different implementations?

Reply:
So in ICS-020, these will be called with an `uint256`.
So I guess there is an implicit assumption that the bank handles _at least_ uint256.

Reply:
Well, then we know the range of amounts. You could do something similar to: https://github.com/informalsystems/quint/blob/eeac951979a14b3fbba8266b3501857abc6c2a62/examples/solidity/ERC20/erc20.qnt#L23-L27

Reply:
> So I guess there is an implicit assumption that the bank handles _at least_ amounts in 0..2^256-1.

Oh, that's interesting. So even if the amounts are in the range, it does not impose a bound on the balances. Maybe you should make a note about this and refer to the place where ICS20 refers to uint256?

Reply:
Well, we know them for the ICS-020 calling code. So I think this sort of logic belongs there, not in the bank module.

Reply:
Aaah, well, but nothing says the bank cannot operate over arbitrary integers.
I guess we can only document this assumption (or, do two bank specs, one with bounded and one with unbounded ints).

Reply:
Also I guess all the amounts are implicitly assumed to be non-negative? Might be good to also note that somewhere, as the minimal logic doesn't capture this.

Reply:
There's a comment in pseudocode in ICS-020 saying `TransferCoins` fails with insufficient balance, so I guess yes. OTOH, there is no such comment for `BurnCoins`...

Reply:
Checking for sufficient balances in [d5fee40](https://github.com/informalsystems/quint/pull/784/commits/d5fee40b52cd97d3853a63100b0b03cde2826360).

Reply:
re: Igor's comments, I've added a README to document any assumptions we make: [9411f5e](https://github.com/informalsystems/quint/pull/784/commits/9411f5e705ae56dcb06c41cf4dcfc6de7bdadbb4)